File: backend/app/routers/correcao.py
"""Router FastAPI para correção monetária IPCA."""

from __future__ import annotations

from fastapi import APIRouter, Query
from pydantic import BaseModel, Field

from app.services.correcao_monetaria import CorrecaoMonetariaService
from app.services.ibge_service import IBGEService

router = APIRouter(prefix="/api/v1/correcao", tags=["correcao"])

_ibge = IBGEService()
_correcao = CorrecaoMonetariaService(_ibge)


class CorrecaoPrecoRequest(BaseModel):
    """Body para correção de preço."""

    valor: float = Field(..., gt=0, description="Valor a corrigir")
    data_origem: str = Field(..., description="Data de origem (YYYY-MM-DD)")
    data_destino: str = Field(..., description="Data de destino (YYYY-MM-DD)")


@router.get("/ipca", summary="Série histórica IPCA")
def serie_ipca(
    ano_inicio: int = Query(2020, ge=2020, description="Ano inicial"),
    ano_fim: int = Query(2026, le=2030, description="Ano final"),
) -> dict:
    """Retorna série histórica do IPCA entre os anos especificados."""
    serie = _ibge.get_serie(ano_inicio, ano_fim)
    return {
        "indice": "IPCA",
        "periodo": {"ano_inicio": ano_inicio, "ano_fim": ano_fim},
        "total_meses": len(serie),
        "dados": [
            {
                "ano": s.ano,
                "mes": s.mes,
                "variacao_mensal": s.variacao_mensal,
                "variacao_acumulada_ano": s.variacao_acumulada_ano,
                "indice_acumulado": s.indice_acumulado,
            }
            for s in serie
        ],
    }


@router.get("/fator", summary="Fator de correção entre duas datas")
def fator_correcao(
    data_inicio: str = Query(..., description="Data inicial (YYYY-MM-DD)"),
    data_fim: str = Query(..., description="Data final (YYYY-MM-DD)"),
) -> dict:
    """Calcula fator de correção IPCA entre duas datas."""
    fator = _correcao.fator_correcao(data_inicio, data_fim)
    variacao = _correcao.variacao_periodo(data_inicio, data_fim)
    return {
        "fator": round(fator, 6),
        "variacao_percentual": variacao,
        "data_inicio": data_inicio,
        "data_fim": data_fim,
        "indice": "IPCA",
    }


@router.post("/preco", summary="Corrigir preço pelo IPCA")
def corrigir_preco(body: CorrecaoPrecoRequest) -> dict:
    """Corrige um preço pela inflação IPCA entre duas datas."""
    return _correcao.corrigir_preco_detalhado(
        body.valor, body.data_origem, body.data_destino
    )


@router.get("/sincronizar", summary="Sincronizar dados IPCA")
def sincronizar() -> dict:
    """Sincroniza dados IPCA com a fonte (IBGE SIDRA)."""
    return _ibge.sincronizar()

File: frontend/src/api/correcao.ts
/**
 * Cliente API para endpoints de correção monetária IPCA.
 */

const API_BASE = (import.meta.env.VITE_API_URL ?? "") + "/api/v1";

export interface DadoIPCA {
  ano: number;
  mes: number;
  variacao_mensal: number;
  variacao_acumulada_ano: number;
  indice_acumulado: number;
}

export interface SerieIPCAResponse {
  indice: string;
  periodo: { ano_inicio: number; ano_fim: number };
  total_meses: number;
  dados: DadoIPCA[];
}

export interface FatorCorrecaoResponse {
  fator: number;
  variacao_percentual: number;
  data_inicio: string;
  data_fim: string;
  indice: string;
}

export interface CorrecaoPrecoResponse {
  valor_original: number;
  valor_corrigido: number;
  fator: number;
  variacao_percentual: number;
  data_origem: string;
  data_destino: string;
  indice: string;
}

async function fetchJSON<T>(url: string, params?: Record<string, unknown>): Promise<T> {
  const urlObj = new URL(url, window.location.origin);
  if (params) {
    for (const [key, value] of Object.entries(params)) {
      if (value !== undefined && value !== null && value !== "") {
        urlObj.searchParams.set(key, String(value));
      }
    }
  }
  const resp = await fetch(urlObj.toString());
  if (!resp.ok) {
    throw new Error(`HTTP ${resp.status}: ${resp.statusText}`);
  }
  return resp.json();
}

export const correcaoAPI = {
  /** Série histórica IPCA. */
  serieIPCA: (anoInicio?: number, anoFim?: number) =>
    fetchJSON<SerieIPCAResponse>(`${API_BASE}/correcao/ipca`, {
      ano_inicio: anoInicio,
      ano_fim: anoFim,
    }),

  /** Fator de correção entre duas datas. */
  fatorCorrecao: (dataInicio: string, dataFim: string) =>
    fetchJSON<FatorCorrecaoResponse>(`${API_BASE}/correcao/fator`, {
      data_inicio: dataInicio,
      data_fim: dataFim,
    }),
};
